core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -174,80 +187,96 @@ public void run() {
      * Periodically invoked to keep track of the load.
      * Launches additional nodes if necessary.
      */
-    private synchronized void update() {
-        Jenkins jenkins = Jenkins.getInstance();
-        lastSuggestedReview = System.currentTimeMillis();
+    private void update() {
+        provisioningLock.lock();
+        try {
+            lastSuggestedReview = System.currentTimeMillis();
+            Queue.withLock(new Runnable() {
```

Comment:
Should document the lock acquisition order (e.g., as Javadoc on `provisioningLock`) lest the next person to edit this introduces a deadlock.